File: backend/services/AuthService.js
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");

const User = require("../models/User");
const AppError = require("../utils/AppError");

// Register Service
const registerUser = async (fullName, email, password, role = "manager") => {

  // Check if email already exists
  const existingUser = await User.findUserByEmail(email);

  if (existingUser) {
    throw new AppError(
    "Email already registered",
    400
       ); 
  }

  // Hash password
  const hashedPassword = await bcrypt.hash(password, 10);

  // Save user
  await User.createUser(fullName, email, hashedPassword, role);

  return {
    message: "User Registered Successfully",
  };
};

// Login Service
const loginUser = async (email, password) => {

  const user = await User.findUserByEmail(email);

  if (!user) {
    throw new AppError(
    "Invalid Email or Password",
    401
     );
  }

  const isMatch = await bcrypt.compare(password, user.password);

  if (!isMatch) {
    throw new AppError(
    "Invalid Email or Password",
    401
      );
  }

  const token = jwt.sign(
    {
      id: user.id,
      email: user.email,
      role: user.role,
    },
    process.env.JWT_SECRET,
    {
      expiresIn: "1d",
    }
  );

  return {
    message: "Login Successful",
    token,
    user: {
      id: user.id,
      fullName: user.full_name,
      email: user.email,
      role: user.role,
    },
  };
};

module.exports = {
  registerUser,
  loginUser,
};

File: backend/src/main/java/com/procureease/backend/service/AuthService.java
package com.procureease.backend.service;

import com.procureease.backend.dto.LoginRequest;
import com.procureease.backend.dto.LoginResponse;
import com.procureease.backend.dto.RegisterRequest;
import com.procureease.backend.dto.UserResponse;
import com.procureease.backend.entity.User;
import com.procureease.backend.exception.ResourceAlreadyExistsException;
import com.procureease.backend.exception.ResourceNotFoundException;
import com.procureease.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;

    private final PasswordEncoder passwordEncoder;

    private final JwtService jwtService;

    // ============================
    // Register User
    // ============================

    public String register(RegisterRequest request) {

        // Check if email already exists
        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new ResourceAlreadyExistsException(
                    "Email already registered"
            );
        }

        // Create User Object
        User user = User.builder()
                .fullName(request.getFullName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(request.getRole())
                .build();

        // Save User
        userRepository.save(user);

        return "User Registered Successfully";
    }

    // ============================
    // Login User
    // ============================

    public LoginResponse login(LoginRequest request) {

        // Find User
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() ->
                        new ResourceNotFoundException("Invalid Email or Password"));

        // Check Password
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new ResourceNotFoundException("Invalid Email or Password");
        }

        // Generate JWT Token
        String token = jwtService.generateToken(user);

        // Build User Response
        UserResponse userResponse = UserResponse.builder()
                .id(user.getId())
                .fullName(user.getFullName())
                .email(user.getEmail())
                .role(user.getRole())
                .build();

        // Return Login Response
        return LoginResponse.builder()
                .message("Login Successful")
                .token(token)
                .user(userResponse)
                .build();
    }
}